src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Composition;
+using System.Runtime.InteropServices.WindowsRuntime;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    [Export(typeof(RequestExecutionQueue)), Shared]
+    internal partial class RequestExecutionQueue
+    {
+        private readonly AsyncQueue<QueueItem> _queue = new AsyncQueue<QueueItem>();
+        private readonly ILspSolutionProvider _solutionProvider;
+        private Solution? _lastMutatedSolution;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RequestExecutionQueue(ILspSolutionProvider solutionProvider)
+        {
+            _solutionProvider = solutionProvider;
+
+            // Start the queue processing
+            _ = ProcessQueueAsync();
```

Comment:
this should be tied to somethign that shutsdown the queue processing.  Sam has an interface fo rthis i believe.

Reply:
Perhaps this could be tied to the LSP shutdown messages.  Just need to make sure that it can be started again if the LSP client decides to restart.

Reply:
I've added a cancellation token for this for now, will add the rest to the TODO list as the lifetime of this class is in question anyway.